### PR TITLE
Add api endpoints

### PR DESCRIPTION
**Pull request by jegorami**
Add endpoints for assignment and section data.
Create tests for api views

Refs TK-29

### CODE REVIEW

File: assignments/tests/conftest.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    open_text_task = OpenTextTaskFactory(section=section_1)
+    budgeting_task = BudgetingTaskFactory(section=section_2)
+
+
+@pytest.fixture
+def assignments_url():
+    return reverse('assignments:assignment-list')
+
+
+
```
Comment (JsseL) [inline assignments/tests/conftest.py]: Some unnecessary blank lines. Files should end with blank line, but not with multiple blank lines.

File: assignments/tests/factories.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+import factory.fuzzy
+from django.utils.text import slugify
+from assignments import models
+from psycopg2.extras import NumericRange
+
+
+class AssignmentFactory(factory.DjangoModelFactory):
+    class Meta:
+        model = models.Assignment
+    name = factory.Sequence(lambda n: 'assignment_%s' % n)
```
Comment (JsseL) [inline assignments/tests/factories.py]: Leave a blank line between Meta class at attributes and class attributes.

File: assignments/tests/factories.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    question = factory.fuzzy.FuzzyText()
+    section = factory.SubFactory(SectionFactory)
+
+
+class BudgetingTargetFactory(factory.DjangoModelFactory):
+    class Meta:
+        model = models.BudgetingTarget
+    name = factory.Sequence(lambda n: 'target_%s' % n)
+    unit_price = factory.fuzzy.FuzzyDecimal(2, 15, precision=2)
+    range = NumericRange(0, 10, '[]')
```
Comment (JsseL) [inline assignments/tests/factories.py]: Didn't you change the range into min and max fields?

File: assignments/tests/factories.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+
+class BudgetingTaskFactory(factory.DjangoModelFactory):
+    class Meta:
+        model = models.BudgetingTask
+    section = factory.SubFactory(SectionFactory)
+    name = factory.Sequence(lambda n: 'budgetingtask_%s' % n)
+    amount_of_consumption = factory.fuzzy.FuzzyDecimal(2, 100, precision=2)
+    unit = factory.Iterator([models.BudgetingTask.UNIT_HA, models.BudgetingTask.UNIT_PCS])
+
```
Comment (JsseL) [inline assignments/tests/factories.py:54]: Unnecessary blank line.

File: assignments/tests/test_views.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
+import pytest
+from assignments.models import Assignment, OpenTextTask, BudgetingTask, BudgetingTarget
+from rest_framework.test import APIClient
+from django.shortcuts import reverse
+from decimal import Decimal
+
+
+class TestApi:
+    @pytest.mark.django_db
+    def test_get_only_opened_assignments(self, create_assignments, assignments_url):
```
Comment (JsseL) [inline assignments/tests/test_views.py]: You could make the test names a bit more descriptive e.g. `test_assignment_endpoint_returns_only_open_assignments`. Descriptive test names gives you really self explaining console output after test runs. Now I had to read the test to understand what it actually tested.

File: assignments/tests/test_views.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        assert Assignment.objects.count() == 2
+        api_client = APIClient()
+        response = api_client.get(assignments_url)
+        response_data = response.json()
+        assert len(response_data) == 1
+        assert response_data[0]['status'] == Assignment.STATUS_OPEN
+
+    @pytest.mark.django_db
+    def test_assignment_sections(self, create_assignments, assignments_url):
```
Comment (JsseL) [inline assignments/tests/test_views.py]: For example this doesn't tell pretty much at all what it is testing. Or it does, but in really high level.

File: assignments/urls.py [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+router = ExtendedSimpleRouter()
+(
+    router.register(r'assignments', AssignmentViewSet, base_name='assignment')
+          .register(r'sections', SectionsViewSet, base_name='section', parents_query_lookups=['assignment__slug'])
+)
+ajax_urls = router.urls
+
+urlpatterns = [
+    url(r'^assignments/(?P<slug>[-\w]+)/$', TemplateView.as_view(template_name='assignments/index.html'), name='index'),
```
Comment (JsseL) [inline assignments/urls.py]: Why do we need a templateview?

File: assignments/urls.py [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+    url(r'^ajax/', include(ajax_urls))
```
Comment (JsseL) [inline assignments/urls.py]: I'd use `api/` since isn't that what it is?

File: assignments/tests/test_views.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+            response_data = response.json()
+            found = False
+            for data_open_text_task in response_data['opentexttasks']:
+                if task.question == data_open_text_task['question']:
+                    found = True
+                    break
+            assert found, 'Cannot find open text task {} in the response'.format(task.question)
+
+    @pytest.mark.django_db
+    def test_assignment_budget_text_tasks(self, create_assignments, assignments_url):
```
Comment (JsseL) [inline assignments/tests/test_views.py]: Again, the name of the test doesn't tell much and the test itself actually does pretty much. Is it really necessary to do all this stuff in one test? Could you split this into multiple more specific test cases?

**Comment by jegorami** [comment]
@JsseL Hope everything is fixed now

File: assignments/tests/test_views.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    def test_all_open_text_tasks_present_in_section_for_open_assignment(self, create_assignments, assignments_url):
+        assignment = Assignment.objects.get(status=Assignment.STATUS_OPEN)
+        api_client = APIClient()
+        open_text_tasks = OpenTextTask.objects.filter(section__assignment=assignment)
+        for task in open_text_tasks:
+            section_url = reverse('assignments:section-detail', args=[assignment.slug, task.section.id])
+            response = api_client.get(section_url)
+            response_data = response.json()
+            found = False
+            for data_open_text_task in response_data['opentexttasks']:
```
Comment (JsseL) [inline assignments/tests/test_views.py:38]: `opentexttasks` sounds like a bad key name in data. Shouldn't it be in snake_case or in camelCase?

**Review by JsseL — COMMENTED** [review]
You could try to keep unrelated changes in separate commits. There are for example some import order changes in `assignments/models.py` that are not related to the rest of the changes. Not a biggie though, but keep it in mind.

Also the tests still seem to be unnecessary complex. You could try to just create minimal amount of necessary objects into the DB and then assert some more specific stuff. Again, doesn't require changes for now. We can move on with these tests. :+1: For having the tests!

Good job. @arvileino your turn :smiley:

File: assignments/tests/test_views.py (hunk 10 above)
Comment (jegorami) [reply assignments/tests/test_views.py:38]: The problem I had there is the fact we have BaseTask abstract model with section field, and I wanted to define related_name for models that inherit BaseTask. I solved it using related_name='%(class)ss' (https://docs.djangoproject.com/en/1.11/topics/db/models/#be-careful-with-related-name-and-related-query-name), but then we are getting related name in the way as is. Do you have better idea?

Comment (jegorami) [reply assignments/tests/test_views.py:38]: I think I could set different name in Serializer data, and set source to be opentexttasks. Is that better?

Comment (JsseL) [reply assignments/tests/test_views.py:38]: If we have to accept such related_name it's at least better to hide such a bad name from the API users.

**Pull request merged**